src/type_checker/checker.rs

```diff
@@ -587,7 +587,7 @@ impl<B: Backend> TypeChecker<B> {
                 return Err(self.error(ErrorKind::NoReturnExpected, stmts.last().unwrap().span))
             }
             (Some(expected), Some(observed)) => {
-                if !observed.match_expected(&expected.kind) {
+                if !observed.match_expected(&expected.kind, true) {
```

Comment:
This allows constant because the value being returned could be either constant or not. For example:
```rust
return arr[idx]; // the element behind the idx could be either a field or a constant field.
```